notebooks_community/clf_constrained_bo.ipynb

```diff
@@ -0,0 +1,840 @@
+    "\n",
+    "[1]: [Wang, F., Parhizkar, M., Harker, A., & Edirisinghe, M. (2024). Constrained composite Bayesian optimization for rational synthesis of polymeric particles. ArXiv, abs/2411.10471](https://arxiv.org/abs/2411.10471).\n",
+    "\n",
+    "[2]: [Tian, Y., Zuniga, A., Zhang, X., Dürholt, J.P., Das, P., Chen, J., Matusik, W., & Konakovic-Lukovic, M. (2024). Boundary Exploration for Bayesian Optimization With Unknown Physical Constraints. ArXiv, abs/2402.07692](https://arxiv.org/abs/2402.07692).\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Set dtype and device"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import torch\n",
+    "\n",
+    "tkwargs = {\n",
+    "    \"dtype\": torch.double,\n",
+    "    \"device\": torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\"),\n",
+    "}\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Problem setup"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We begin with setting up a classical synthetic problem `Townsend` presented in [3]. The objective to be maximized is:\n",
+    "\n",
+    "$$\n",
+    "\\begin{align*}\n",
+    "y_{\\text{obj}} &= \\cos\\left((x_1 - 0.1) x_2\\right)^2 + x_1 \\sin(3 x_1 + x_2)\n",
+    "\\end{align*}\n",
+    "$$\n",
+    "\n",
+    "Subjected to the constraint:\n",
+    "$$\n",
+    "y_{\\text{con}} = \\left(\\left(2 \\cos(t) - 0.5 \\cos(2t) - 0.25 \\cos(3t) - 0.125 \\cos(4t)\\right)^2 + \\left(2 \\sin(t)\\right)^2 - x_1^2 - x_2^2\\right)>0\n",
+    "$$ \n",
+    "where $t = \\arctan\\left(\\frac{x_1}{x_2}\\right)$\n",
+    "\n",
+    "Here, we follow a natural representation where $y_{\\text{con}}=1$ indicates a feasible condition. We will train a classification model to predict the feasibility of the point. Note that in BoTorch's implementation, **negative values** indicate feasibility, thus we need to do conversion later when feeding feasibility into the pipeline.\n",
+    "\n",
+    "\n",
+    "[3]: Townsend, A. (2014). Constrained optimization in Chebfun. https://www.chebfun.org/examples/opt/ConstrainedOptimization.html"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Townsend():\n",
+    "    def __init__(self):\n",
+    "        self.dim = 2\n",
+    "        self.lb = torch.tensor([-2.25, 2.25], **tkwargs)\n",
+    "        self.ub = torch.tensor([-2.5, 1.75], **tkwargs)\n",
+    "        self.bounds = torch.stack([self.lb, self.ub]).T\n",
+    "        self._optimal_value = 1.660\n",
+    "        self.name = \"Townsend\"\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return self.objective(x)\n",
+    "\n",
+    "    def constraint(self, x):\n",
```

Comment:
Maybe calling this `is_feasible()` is more intuitive in that it it makes it clear this is a boolean observation?